From 8c3a33de08616ce6199e37bc6d3066f6f92c29b1 Mon Sep 17 00:00:00 2001
From: Herb <herb@3speedhub.com>
Date: Tue, 22 Apr 2025 14:09:08 -0400
Subject: [PATCH 1/8] Add new search method and deprecate old one

---
 src/Issue/IssueService.php | 39 +++++++++++++++++++++++++++++++++++++-
 1 file changed, 38 insertions(+), 1 deletion(-)

diff --git a/src/Issue/IssueService.php b/src/Issue/IssueService.php
index d0da533..7cd02d8 100644
--- a/src/Issue/IssueService.php
+++ b/src/Issue/IssueService.php
@@ -481,6 +481,43 @@ public function transition(string|int $issueIdOrKey, Transition $transition): ?s
      * Search issues.
      *
      * @param string $jql
+     * @param string $nextPageToken
+     * @param int    $maxResults
+     * @param array  $fields
+     * @param array  $expand
+     * @param array  $reconcileIssues
+     *
+     * @throws \JsonMapper_Exception
+     * @throws JiraException
+     *
+     * @return IssueSearchResult
+     */
+    public function search(string $jql, string $nextPageToken = '', int $maxResults = 50, array $fields = [], array $expand = [], array $reconcileIssues = []): IssueSearchResult
+    {
+        $data = json_encode([
+            'jql'             => $jql,
+            'nextPageToken'   => $nextPageToken,
+            'maxResults'      => $maxResults,
+            'fields'          => $fields,
+            'expand'          => $expand,
+            'reconcileIssues' => $reconcileIssues,
+        ]);
+
+        $ret = $this->exec('search/jql', $data, 'POST');
+        $json = json_decode($ret);
+
+        $result = $this->json_mapper->map(
+            $json,
+            new IssueSearchResult()
+        );
+
+        return $result;
+    }
+
+    /**
+     * Search issues (old).
+     *
+     * @param string $jql
      * @param int    $startAt
      * @param int    $maxResults
      * @param array  $fields
@@ -492,7 +529,7 @@ public function transition(string|int $issueIdOrKey, Transition $transition): ?s
      *
      * @return IssueSearchResult
      */
-    public function search(string $jql, int $startAt = 0, int $maxResults = 15, array $fields = [], array $expand = [], bool $validateQuery = true): IssueSearchResult
+    public function search_old(string $jql, int $startAt = 0, int $maxResults = 15, array $fields = [], array $expand = [], bool $validateQuery = true): IssueSearchResult
     {
         $data = json_encode([
             'jql'           => $jql,

From 6bf95c65550dce7e48ee1be173b523cb02a4faa6 Mon Sep 17 00:00:00 2001
From: Herb <herb@3speedhub.com>
Date: Tue, 22 Apr 2025 21:50:24 -0400
Subject: [PATCH 2/8] Update IssueService.php

---
 src/Issue/IssueService.php | 50 ++++++--------------------------------
 1 file changed, 8 insertions(+), 42 deletions(-)

diff --git a/src/Issue/IssueService.php b/src/Issue/IssueService.php
index 7cd02d8..48c2348 100644
--- a/src/Issue/IssueService.php
+++ b/src/Issue/IssueService.php
@@ -484,7 +484,7 @@ public function transition(string|int $issueIdOrKey, Transition $transition): ?s
      * @param string $nextPageToken
      * @param int    $maxResults
      * @param array  $fields
-     * @param array  $expand
+     * @param string  $expand
      * @param array  $reconcileIssues
      *
      * @throws \JsonMapper_Exception
@@ -492,55 +492,21 @@ public function transition(string|int $issueIdOrKey, Transition $transition): ?s
      *
      * @return IssueSearchResult
      */
-    public function search(string $jql, string $nextPageToken = '', int $maxResults = 50, array $fields = [], array $expand = [], array $reconcileIssues = []): IssueSearchResult
+    public function search(string $jql, string $nextPageToken = '', int $maxResults = 50, array $fields = [], string $expand = '', array $reconcileIssues = []): IssueSearchResult
     {
-        $data = json_encode([
+        $data = [
             'jql'             => $jql,
-            'nextPageToken'   => $nextPageToken,
             'maxResults'      => $maxResults,
             'fields'          => $fields,
             'expand'          => $expand,
             'reconcileIssues' => $reconcileIssues,
-        ]);
-
-        $ret = $this->exec('search/jql', $data, 'POST');
-        $json = json_decode($ret);
-
-        $result = $this->json_mapper->map(
-            $json,
-            new IssueSearchResult()
-        );
-
-        return $result;
-    }
+        ];
 
-    /**
-     * Search issues (old).
-     *
-     * @param string $jql
-     * @param int    $startAt
-     * @param int    $maxResults
-     * @param array  $fields
-     * @param array  $expand
-     * @param bool   $validateQuery
-     *
-     * @throws \JsonMapper_Exception
-     * @throws JiraException
-     *
-     * @return IssueSearchResult
-     */
-    public function search_old(string $jql, int $startAt = 0, int $maxResults = 15, array $fields = [], array $expand = [], bool $validateQuery = true): IssueSearchResult
-    {
-        $data = json_encode([
-            'jql'           => $jql,
-            'startAt'       => $startAt,
-            'maxResults'    => $maxResults,
-            'fields'        => $fields,
-            'expand'        => $expand,
-            'validateQuery' => $validateQuery,
-        ]);
+        if ($nextPageToken) {
+            $data['nextPageToken'] = $nextPageToken;
+        }
 
-        $ret = $this->exec('search', $data, 'POST');
+        $ret = $this->exec('search/jql', json_encode($data), 'POST');
         $json = json_decode($ret);
 
         $result = $this->json_mapper->map(

From 3899df36b9687ada1114ecd5cf8df5d5a3999d8e Mon Sep 17 00:00:00 2001
From: Herb <herb@3speedhub.com>
Date: Tue, 22 Apr 2025 21:53:11 -0400
Subject: [PATCH 3/8] Update IssueService.php

---
 src/Issue/IssueService.php | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/Issue/IssueService.php b/src/Issue/IssueService.php
index 48c2348..63f0719 100644
--- a/src/Issue/IssueService.php
+++ b/src/Issue/IssueService.php
@@ -506,7 +506,7 @@ public function search(string $jql, string $nextPageToken = '', int $maxResults
             $data['nextPageToken'] = $nextPageToken;
         }
 
-        $ret = $this->exec('search/jql', json_encode($data), 'POST');
+        $ret = $this->exec('search//jql', json_encode($data), 'POST');
         $json = json_decode($ret);
 
         $result = $this->json_mapper->map(

From 7256a6e33d9a83bce92f189296b491b015c55341 Mon Sep 17 00:00:00 2001
From: Herb <herb@3speedhub.com>
Date: Tue, 22 Apr 2025 21:54:54 -0400
Subject: [PATCH 4/8] style fix

---
 src/Issue/IssueService.php | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/Issue/IssueService.php b/src/Issue/IssueService.php
index 63f0719..20781a2 100644
--- a/src/Issue/IssueService.php
+++ b/src/Issue/IssueService.php
@@ -484,7 +484,7 @@ public function transition(string|int $issueIdOrKey, Transition $transition): ?s
      * @param string $nextPageToken
      * @param int    $maxResults
      * @param array  $fields
-     * @param string  $expand
+     * @param string $expand
      * @param array  $reconcileIssues
      *
      * @throws \JsonMapper_Exception

From 999a46dbd2a1e1ba4f03c82b455430bca9b29f33 Mon Sep 17 00:00:00 2001
From: Herb <herb@3speedhub.com>
Date: Tue, 22 Apr 2025 22:35:33 -0400
Subject: [PATCH 5/8] Update IssueSearchResult.php

---
 src/Issue/IssueSearchResult.php | 77 +++------------------------------
 1 file changed, 7 insertions(+), 70 deletions(-)

diff --git a/src/Issue/IssueSearchResult.php b/src/Issue/IssueSearchResult.php
index bae70a3..50a0bde 100644
--- a/src/Issue/IssueSearchResult.php
+++ b/src/Issue/IssueSearchResult.php
@@ -10,22 +10,7 @@ class IssueSearchResult
     /**
      * @var string
      */
-    public $expand;
-
-    /**
-     * @var int
-     */
-    public $startAt;
-
-    /**
-     * @var int
-     */
-    public $maxResults;
-
-    /**
-     * @var int
-     */
-    public $total;
+    public $nextPageToken;
 
     /**
      * @var \JiraCloud\Issue\Issue[]
@@ -33,51 +18,19 @@ class IssueSearchResult
     public $issues;
 
     /**
-     * @return int
-     */
-    public function getStartAt()
-    {
-        return $this->startAt;
-    }
-
-    /**
-     * @param int $startAt
-     */
-    public function setStartAt($startAt)
-    {
-        $this->startAt = $startAt;
-    }
-
-    /**
-     * @return int
-     */
-    public function getMaxResults()
-    {
-        return $this->maxResults;
-    }
-
-    /**
-     * @param int $maxResults
-     */
-    public function setMaxResults($maxResults)
-    {
-        $this->maxResults = $maxResults;
-    }
-
-    /**
-     * @return int
+     * @return string
      */
-    public function getTotal()
+    public function getNextPageToken()
     {
-        return $this->total;
+        return $this->nextPageToken;
     }
 
     /**
-     * @param int $total
+     * @param string $nextPageToken
      */
-    public function setTotal($total)
+    public function setNextPageToken($nextPageToken)
     {
-        $this->total = $total;
+        $this->nextPageToken = $nextPageToken;
     }
 
     /**
@@ -105,20 +58,4 @@ public function getIssue($ndx)
     {
         return $this->issues[$ndx];
     }
-
-    /**
-     * @return string
-     */
-    public function getExpand()
-    {
-        return $this->expand;
-    }
-
-    /**
-     * @param string $expand
-     */
-    public function setExpand($expand)
-    {
-        $this->expand = $expand;
-    }
 }

From 6fdd6d5dbaf3468c5bad65df490bd735555d35d8 Mon Sep 17 00:00:00 2001
From: Herb <herb@3speedhub.com>
Date: Tue, 22 Apr 2025 23:59:29 -0400
Subject: [PATCH 6/8] Create IssueBulkResult.php

---
 src/Issue/IssueBulkResult.php | 82 +++++++++++++++++++++++++++++++++++
 1 file changed, 82 insertions(+)
 create mode 100644 src/Issue/IssueBulkResult.php

diff --git a/src/Issue/IssueBulkResult.php b/src/Issue/IssueBulkResult.php
new file mode 100644
index 0000000..cfe3279
--- /dev/null
+++ b/src/Issue/IssueBulkResult.php
@@ -0,0 +1,82 @@
+<?php
+
+namespace JiraCloud\Issue;
+
+/**
+ * Issue search result.
+ */
+class IssueBulkResult
+{
+    /**
+     * @var string
+     */
+    public $expand;
+
+    /**
+     * @var \JiraCloud\Issue\Issue[]
+     */
+    public $issues;
+
+    /**
+     * @var array
+     */
+    public $issueErrors;
+
+    /**
+     * @return array
+     */
+    public function getIssueErrors()
+    {
+        return $this->issueErrors;
+    }
+
+    /**
+     * @param array $issueErrors
+     */
+    public function setIssueErrors($issueErrors)
+    {
+        $this->issueErrors = $issueErrors;
+    }
+
+    /**
+     * @return Issue[]
+     */
+    public function getIssues()
+    {
+        return $this->issues;
+    }
+
+    /**
+     * @param Issue[] $issues
+     */
+    public function setIssues($issues)
+    {
+        $this->issues = $issues;
+    }
+
+    /**
+     * @param int $ndx
+     *
+     * @return Issue
+     */
+    public function getIssue($ndx)
+    {
+        return $this->issues[$ndx];
+    }
+
+    /**
+     * @return string
+     */
+    public function getExpand()
+    {
+        return $this->expand;
+    }
+
+    /**
+     * @param string $expand
+     */
+    public function setExpand($expand)
+    {
+        $this->expand = $expand;
+    }
+}

From bfb3c4e2de6eddf51d8d3eac9c9dcbe3244ed21a Mon Sep 17 00:00:00 2001
From: Herb <herb@3speedhub.com>
Date: Wed, 23 Apr 2025 00:00:46 -0400
Subject: [PATCH 7/8] Update IssueService.php

---
 src/Issue/IssueService.php | 57 ++++++++++++++++++++++++++++++++++++++
 1 file changed, 57 insertions(+)

diff --git a/src/Issue/IssueService.php b/src/Issue/IssueService.php
index 20781a2..6832b27 100644
--- a/src/Issue/IssueService.php
+++ b/src/Issue/IssueService.php
@@ -517,6 +517,63 @@ public function search(string $jql, string $nextPageToken = '', int $maxResults
         return $result;
     }
 
+    /**
+     * Search issues.
+     *
+     * @param string $jql
+     *
+     * @throws \JsonMapper_Exception
+     * @throws JiraException
+     *
+     * @return string[] array of count
+     *
+     * @phpstan-return array<string>
+     */
+    public function searchApproximateCount(string $jql): array
+    {
+        $data = json_encode([
+            'jql' => $jql,
+        ]);
+
+        $ret = $this->exec('search//approximate-count', $data, 'POST');
+
+        return json_decode($ret, true);
+    }
+
+    /**
+     * Bulk fetch issues.
+     *
+     * @param array $issueIdsOrKeys
+     * @param array $fields
+     * @param array $expand
+     * @param bool  $fieldsByKeys
+     *
+     * @throws \JsonMapper_Exception
+     * @throws JiraException
+     *
+     * @return IssueBulkResult
+     */
+    public function bulkFetch(array $issueIdsOrKeys, array $fields = [], array $expand = [], bool $fieldsByKeys = false): IssueBulkResult
+    {
+        $data = json_encode([
+            'issueIdsOrKeys'  => $issueIdsOrKeys,
+            'fields'          => $fields,
+            'expand'          => $expand,
+            'fieldsByKeys' => $fieldsByKeys,
+        ]);
+
+        $ret = $this->exec('issue//bulkfetch', $data, 'POST');
+
+        $json = json_decode($ret);
+
+        $result = $this->json_mapper->map(
+            $json,
+            new IssueBulkResult()
+        );
+
+        return $result;
+    }
+
     /**
      * get TimeTracking info.
      *

From b775f09ef399d410e61388aa06ef3832781ebffc Mon Sep 17 00:00:00 2001
From: Herb <herb@3speedhub.com>
Date: Wed, 23 Apr 2025 00:01:56 -0400
Subject: [PATCH 8/8] Update IssueService.php

---
 src/Issue/IssueService.php | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/Issue/IssueService.php b/src/Issue/IssueService.php
index 6832b27..71a8496 100644
--- a/src/Issue/IssueService.php
+++ b/src/Issue/IssueService.php
@@ -559,7 +559,7 @@ public function bulkFetch(array $issueIdsOrKeys, array $fields = [], array $expa
             'issueIdsOrKeys'  => $issueIdsOrKeys,
             'fields'          => $fields,
             'expand'          => $expand,
-            'fieldsByKeys' => $fieldsByKeys,
+            'fieldsByKeys'    => $fieldsByKeys,
         ]);
 
         $ret = $this->exec('issue//bulkfetch', $data, 'POST');